vimeo/VimeoNetworking
# VIM-7293: Changes default JSON serializer behavior on VimeoSessi…

#### Ticket

[VIM-7293](https://vimean.atlassian.net/browse/VIM-7293)

#### Pull Request Checklist


#### Issue Summary

Currently the `VimeoSessionManager` that inherits from `AFHTTPURLSessionManager` and handles requests on VimeoClient's behalf is tightly coupled with both the request and response serializers provided by the library.

Basically every time a request is made the data returned with the response is automatically serialized by the internal of `AFNetworking` and returned to us as a JSON object, and there is no way to opt-out of this behavior.

The goal of this ticket is to make the required changes so that we can opt-in to JSON serialization of responses as appropriate, offering the ability to receive raw data when needed (effectively bypassing the JSON response serializer), while maintaining backwards compatibility with the existing API and its consumers.

This will also allow us to support `Decodable` types and eventually remove our dependency on the dynamic object mapping.

#### Implementation Summary

The important changes in this PR are inside `VimeoSessionManager`.

Firstly, `VimeoSessionManager` no longer inherits from `AFHTTPSessionManager`. This change is mostly cosmetic at the moment as internally we still use an instance of `AFHTTPSessionManager` to manage requests.

Previously the session manager was initialized with a default JSON request & response serializers (`VimeoRequestSerializer` and `VimeoResponseSerializer`). This was changed so the default AFNetworking serializers are used instead.

What this effectively means is that both requests and responses using `VimeoSessionManager` are now "opt-in" serializable. We can also opt to receive raw `Data` types from the response, allowing us to opt-in to the appropriate serialization depending on the calling site and public API used.

The next change once the above was in place was to add the public APIs to allow the calling site to choose how they want the response to be treated. The current options are:

- Raw `Data`
- JSON serialization using `VimeoResponseSerializer` (previous default behavior)
- Decodable types using `JSONDecoder`
- Uploads & downloads

In addition to the above I have also refactored the public API for better ergonomics and stricter, more meaningful responses using `Result` types.

There were also some cosmetic changes and addition of wrapper methods on `VimeoSessionManager`. The main reason for this is that since we no longer inherit from `AFHTTPSessionManager` I decided to make the underlying session manager internal to the framework and only expose whatever methods the clients (Upload and iOS) already use via the session manager.

#### How to Test

- Build and run, ensure all unit tests pass
- On the client side, build and run [this PR](https://github.vimeows.com/MobileApps/Vimeo-iOS/pull/3091) and ensure the app behaves as expected, including downloading and uploading videos.
